web/src/reaction_notifications.js

```diff
@@ -0,0 +1,141 @@
+import * as message_store from "./message_store";
+import * as people from "./people";
+import {get_local_reaction_id} from "./reactions";
+import * as settings_config from "./settings_config";
+import {current_user} from "./state_data";
+import * as ui_util from "./ui_util";
+import {user_settings} from "./user_settings";
+import * as user_topics from "./user_topics";
+
+function generate_notification_title(emoji_name, user_ids) {
+    const usernames = people.get_display_full_names(
+        user_ids.filter((user_id) => user_id !== current_user.user_id),
+    );
+
+    const current_user_reacted = user_ids.length !== usernames.length;
+
+    const context = {
+        emoji_name: ":" + emoji_name + ":",
+    };
+
+    if (user_ids.length === 1) {
+        context.username = usernames[0];
+        return $t({defaultMessage: "{username} reacted with {emoji_name}."}, context);
+    }
+
+    if (user_ids.length === 2 && current_user_reacted) {
+        context.other_username = usernames[0];
+        return $t(
+            {
+                defaultMessage: "You and {other_username} reacted with {emoji_name}.",
+            },
+            context,
+        );
+    }
+
+    context.total_reactions = usernames.length;
+    context.last_username = usernames.at(-1);
+
+    return $t(
+        {
+            defaultMessage:
+                "{last_username} and {total_reactions} others reacted with {emoji_name}.",
+        },
+        context,
+    );
+}
+
+export function reaction_is_notifiable(message) {
+    if (message.current_user_reacted) {
+        return false;
+    }
+
+    if (!message.sent_by_me) {
+        return false;
+    }
+
+    if (message.type === "private" && user_settings.enable_dm_reactions_notifications) {
+        return true;
+    }
+
+    if (
+        message.type === "stream" &&
+        user_settings.streams_reaction_notification ===
+            settings_config.streams_reaction_notification_values.always.code
+    ) {
+        return true;
+    }
+
+    if (
+        message.type === "stream" &&
+        user_settings.streams_reaction_notification ===
+            settings_config.streams_reaction_notification_values.never.code
+    ) {
+        return false;
+    }
+
+    if (
+        message.type === "stream" &&
+        user_topics.is_topic_followed(message.stream_id, message.topic) &&
+        user_settings.streams_reaction_notification ===
+            settings_config.streams_reaction_notification_values.followed_topics.code
+    ) {
+        return true;
+    }
+
+    if (
+        message.type === "stream" &&
+        !user_topics.is_topic_unmuted(message.stream_id, message.topic) &&
+        user_settings.streams_reaction_notification ===
+            settings_config.streams_reaction_notification_values.unmuted_topics.code
+    ) {
+        return true;
+    }
+
+    // Everything else is on the table; next filter based on notification
+    // settings.
```

Comment:
Sure, I'll clarify the comment. It refers to filtering based on this notification settings.

> > Hm, for some reason it worked at first for me (using two browsers with two different accounts), but then it stopped working. Is there something I have to be careful of to make sure I'm testing it correctly?
> 
> Hmm, are you sure you had the Visual and Audible settings for Stream notifications checked? It could affect how notifications are received, especially when you choose "Always" or "Unmuted topics" from the "Stream Messages" dropdown menu.
> 
> <img alt="Screenshot 2024-03-15 at 9 40 28 PM" width="758" src="https://private-user-images.githubusercontent.com/72064462/313261581-2674547d-b104-4f21-b9be-cc4318dd1f12.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTA4NzU3MTksIm5iZiI6MTcxMDg3NTQxOSwicGF0aCI6Ii83MjA2NDQ2Mi8zMTMyNjE1ODEtMjY3NDU0N2QtYjEwNC00ZjIxLWI5YmUtY2M0MzE4ZGQxZjEyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzE5VDE5MTAxOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM0OTkyNDZlYmQ4ZDBlYWU4MmRmMTNlZTQwYWNjOGI0YTIzZmEzMjQ4MTZlZWM2OTliMjZhNzgyOWI4Nzc1NGUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.-0EOU1OHoT9mhDQmJ5qaivqb6KY4WnrORHwG1ol3_ls">